[최지수] Step3 리뷰 요청합니다~

jisu1211 opened the pull request:
안녕하세요 근환님!
`3단계 인증을 통한 기능 구현` 미션 완료하여 리뷰 요청합니다 ㅎㅎ

미션을 진행하면서 궁금한 점이 2가지 생겼는데 근환님의 의견이 궁금하여 여쭤보고자 합니다! ㅎㅎ

1. 현재 엔티티 구성에서 다대일 단방향으로 구성된 `Section-Station`의 관계와 `Favorite-Station` 관계, `Favorite-Member` 관계가 있습니다. 이 관계들은 서로 생명주기가 완벽하게 일치하지는 않고 Station, Member에서 Section, Favorite의 정보를 보관하고 있을 필요가 없을 것 같다는 생각에서 이렇게 구성하였습니다! 그런데 Station, Member가 삭제될 때는 삭제되는 Station, Member를 포함하고 있는 Section, Favorite도 삭제되어야 할 것 같은데 여기서 고민이 되는 부분이 있습니다. 
->  `StationService`, `MemberService`에서 수행되는 `delete` 로직에서 삭제되기 전 Section, Favorite을 삭제하는 것
->  다대일 단방향 관계를 다대일 양방향으로 수정해 영속성 전이로 관리
어떤 방법이 더 낫다고 생각하시는지 궁금합니다!

2. 현재 구현된 즐겨찾기 기능에서 딱히 `구간`에 대한 검증을 하고 있지는 않은데 source와 target이 서로 연결되었을 경우에만 즐겨찾기를 추가할 수 있도록 하는 validation을 추가하는 것이 좋을까요? 

이번 리뷰도 잘부탁드립니다!
감사합니다 😊

jisu1211 committed:
refactor: 2단계 피드백 반영
- README.md 인수 조건 예시 추가
- final 키워드 추가
- 매직넘버 상수로 추출

jisu1211 committed:
docs: README.md 수정

jisu1211 committed:
refactor: Member 리팩터링
- Email 별도의 클래스로 분리
- Password 별도의 클래스로 분리
- Member 관련 에러 메시지를 관리하는 MemberExceptionCode 추가

test: EmailTest, MemberTest, PasswordTest 추가

jisu1211 committed:
test: Member 인수 테스트 리팩터링
- MemberAcceptance, MemberAcceptanceTest 클래스 분리
- 인수 테스트 시나리오 추가해서 메서드 분리

jisu1211 committed:
test: Member 인수 테스트 리팩터링

jisu1211 committed:
fix: @Embedded 사용된 컬럼으로 JPQL 검색할 때 발생하는 이슈 수정
test: MemberRepositoryTest 추가

jisu1211 committed:
refactor: Controller, Repository, Service 리팩터링
- @RequestMapping 추가
- @Repository 추가
- @Transactional 추가
- findById, findByEmail 메서드 분리
- final 키워드 추가

jisu1211 committed:
feat: custom exception 추가

test: 로그인한 사용자의 토큰 정보로 수정/삭제/조회하는 인수 테스트 추가

jisu1211 committed:
test: Auth, Member 인수 테스트 케이스 추가

jisu1211 committed:
test: Member 인수 테스트 케이스 추가, MemberServiceTest 추가

jisu1211 committed:
refactor: 사용하지 않는 import문 제거

jisu1211 committed:
feat: 사용자 즐겨찾기 기능 도메인 Favorite 추가
- validation 추가
- 즐겨찾기 관련 에러 메시지를 관리하는 FavoriteExceptionCode 추가

test: Favorite 인수 테스트 패키지 변경, FavoriteTest 추가

jisu1211 committed:
feat: 즐겨찾기 생성 기능 추가
- FavoriteRepository 추가
- FavoriteService 추가
- FavoriteController 추가
- dto(FavoriteRequest, FavoriteResponse) 추가

refactor: final 키워드 추가

test: 즐겨찾기 생성 인수 테스트 추가

jisu1211 committed:
feat: 즐겨찾기 목록 조회 기능 추가
test: 즐겨찾기 목록 조회 인수 테스트 추가, FavoriteRepositoryTest 추가

jisu1211 committed:
feat: 즐겨찾기 삭제 기능 추가
test: 즐겨찾기 삭제 관련 인수 테스트 추가

jisu1211 committed:
test: 즐겨찾기 인수 테스트 케이스 추가, FavoriteServiceTest 추가

jisu1211 committed:
docs: README.md 추가

geunhwanlee commented on src/main/java/nextstep/subway/auth/application/AuthService.java:
```diff
@@ -6,34 +6,43 @@
         this.jwtTokenProvider = jwtTokenProvider;
     }
 
+    @Transactional
```
```suggestion
    @Transactional(readOnly = true)
```
`readOnly=true` 옵션을 붙이는건 어떨까요?

jisu1211 replied:
상세한 답변 감사드립니다 ㅎㅎ 
근환님이 말씀해주신대로 Station이 삭제될때 Section에 대한 처리를 `StationService`에서 하는게 적절한지가
고민이 많이 되었었는데 답변이 많은 도움이 되었습니다! ㅎㅎ
구간에 대한 validation도 추가하지 않는 방향으로 진행하겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/auth/application/AuthService.java:
```diff
@@ -6,34 +6,43 @@
         return new TokenResponse(token);
     }
 
+    @Transactional(readOnly = true)
+    public Member findMemberByEmail(String email) {
+        return memberRepository.findByEmail(email)
+                .orElseThrow(() -> new AuthorizationException(MemberExceptionCode.NOT_FOUND_BY_EMAIL.getMessage()));
+    }
```
`AuthService`에서만 사용중인 메서드인데 `public`으로 만드신 이유가 있을까요? 🤔

jisu1211 replied:
제가 알고있기로는 `@Transactional`이 public 메서드가 아닐때는 동작하지 않는다고 알고 있어서 그렇게 했습니다!

참고 링크: https://stackoverflow.com/questions/4396284/does-spring-transactional-attribute-work-on-a-private-method

geunhwanlee replied:
`@Transactional`은 public 메서드일때 뿐만 아니라 `authService.findMemberByEmail`과 같이 스프링 컴포넌트를 통해 실행되는 메서드에만 동작해요.
`@Transactional`은 스프링 AOP를 통해서 구현되었기 때문이에요!
[참고링크](https://velog.io/@ann0905/AOP와-Transactional의-동작-원리)

이 케이스에서는 `findMemberByEmail`은 `@Transactional` 없이 private로 선언하고
호출하는 메서드에 `@Transactional`을 선언하는게 맞지않을까 생각해요. 😀

jisu1211 replied:
넵 좋은 자료 정말 감사합니다! 
`@Transactional`의 동작에 대해 좀 더 자세히 알 수 있게 되었네요 ㅎㅎ 
`@Transactional`에 대해서도 더 공부해봐야겠습니다~!
근환님이 처음 제안해주신대로 수정하도록 하겠습니다!

geunhwanlee commented on src/main/java/nextstep/subway/auth/application/AuthService.java:
```diff
@@ -6,34 +6,43 @@
     public LoginMember findMemberByToken(String credentials) {
         if (!jwtTokenProvider.validateToken(credentials)) {
-            return new LoginMember();
+            throw new AuthorizationException(MemberExceptionCode.INVALID_TOKEN.getMessage());
```
적절한 예외를 던지도록 수정하셨네요! 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,68 @@
+        Member member = findMemberId(memberId);
+        Station source = findStationById(request.getSource());
+        Station target = findStationById(request.getTarget());
+        Favorite favorite = favoriteRepository.save(request.toFavorite(member, source, target));
```
```suggestion
        Favorite favorite = favoriteRepository.save(new Favorite(member, source, target));
```
`request`랑 관계없이 `Favorite`를 만들고있는데 `Favorite`의 생성자를 통해 만들어도 되지 않을까요?

geunhwanlee commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,68 @@
+        return FavoriteResponse.of(favorite);
+    }
+
+    @Transactional(readOnly = true)
+    public Member findMemberId(Long memberId) {
+        return memberRepository.findById(memberId)
+                .orElseThrow(() -> new NotFoundException(MemberExceptionCode.NOT_FOUND_BY_ID.getMessage()));
+    }
+
+    @Transactional(readOnly = true)
+    public Station findStationById(Long stationId) {
+        return stationRepository.findById(stationId)
+                .orElseThrow(() -> new NotFoundException(StationExceptionCode.NOT_FOUND_BY_ID.getMessage()));
+    }
```
private 메서드로 사용해도 될거같아요!

jisu1211 replied:
위에 public으로 작성한 이유를 댓글에 달아두었습니다~!

geunhwanlee commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,68 @@
+                .collect(toList());
+    }
+
+    @Transactional
+    public void deleteFavorite(Long favoriteId) {
+        favoriteRepository.deleteById(favoriteId);
+    }
```
로그인 사용자가 등록한 즐겨찾기인지 검증할 필요가 있지 않을까요? 🤔

jisu1211 replied:
저는 요청을 보낼때 토큰으로 검증했으니 추가적으로 검증할 필요가 없다고 생각해서 이렇게 작성하게 되었는데 추가적인 검증을 해주는게 더 좋을까요??

geunhwanlee replied:
컨트롤러 레이어에서 검증한것과 별개로
애플리케이션 레이어에서 검증해야하지 않을까 생각해요. 😀

jisu1211 replied:
검증하는 로직 추가하도록 하겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,93 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
+    @JoinColumn(nullable = false)
+    private Station source;
```
`@ManyToOne`의 디폴트 `fetchType`은 즉시로딩이에요.
즉시로딩은 어떤 문제점이 있을까요? 🤔

jisu1211 replied:
즉시 로딩하게 되면 Favorite이 조회될때 Station, Member 데이터도 쿼리가 같이 발생하게 되는데 Favorite을 조회할 때 같이 조회될 일이 많을 것 같아서 즉시 로딩으로 설계해봤습니다! nullable = false로 해서 inner join이 발생하도록 해봤는데 LAZY로 변경하는 것이 더 좋을까요??

geunhwanlee replied:
엔티티의 연관관계 로딩은 항상 지연로딩을 사용하는것이 좋다고 알고있어요! 😀
[참고해보시면 좋을것같아요.](https://girawhale.tistory.com/129)

jisu1211 replied:
좋은 자료 감사합니다~
항상 연관관계 매핑할 때마다 고민되는 지점인 것 같아요 ㅠㅠ 
아직 JPA에 대해 모르는게 많다보니 이런저런 시도를 해보고 있는데 
앞으로는 꼭 필요한 상황이 아니면 지연로딩으로 설정하고 진행하도록 하겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/member/domain/Member.java:
```diff
@@ -1,29 +1,30 @@
-    private String email;
-    private String password;
+
+    @Embedded
+    private Email email;
+
+    @Embedded
+    private Password password;
```
값을 객체로 포장하셨네요 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/member/domain/Member.java:
```diff
@@ -50,8 +47,25 @@ public void update(Member member) {
     public void checkPassword(String password) {
-        if (!StringUtils.equals(this.password, password)) {
-            throw new AuthorizationException();
+        this.password.checkPassword(password);
```
👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/member/ui/MemberController.java:
```diff
@@ -12,51 +11,53 @@
-    @PostMapping("/members")
-    public ResponseEntity createMember(@RequestBody MemberRequest request) {
+    @PostMapping
+    public ResponseEntity<MemberResponse> createMember(@RequestBody MemberRequest request) {
```
API 스펙을 변경하셨는데 이유가 있을까요? 🤔

jisu1211 replied:
리턴되는 타입을 정확하게 명시해주는 것이 좋을 것 같아서 변경해봤습니다!

geunhwanlee replied:
명시한것뿐만 아니라 body도 추가하셔서 여쭤본거였습니다. 😀
괜찮은것같아요! 👍🏻

jisu1211 replied:
다시 보니 제가 질문의 의도를 약간 다르게 파악해서 댓글을 달았네요!
아마 인수 테스트 작성하면서 생성된 Member의 내용을 확인하고 싶어서 저렇게 추가했던 것 같습니다 ㅎㅎ
그리고 다른 인수 테스트에서 Member를 사용할 때도 MemberResponse에서 참조해서 사용할 수 있도록 추가했던 것으로 기억하고 있습니다!

```java
MemberAcceptance.create_member(EMAIL, PASSWORD, 20).as(MemberResponse.class);
```

geunhwanlee commented on src/test/java/nextstep/subway/member/acceptance/ManagingMemberInfoAcceptanceTest.java:
```diff
@@ -0,0 +1,127 @@
+    @Test
+    void updateMyInfo() {
+        // when
+        ExtractableResponse<Response> response = MemberAcceptance.member_was_updated(tokenResponse,
```
개인적인 생각으로는 `member_was_updated`보다는 `update_me`와 같이 내정보를 수정한다는 의미를 담는게 좋지 않을까요?

jisu1211 replied:
말씀해주신 이름이 더 명확해보이네요! 수정하겠습니다 ㅎㅎ

geunhwanlee reviewed (changes requested):
3단계 미션 잘 해주셨어요! 👍🏻

몇가지 생각해볼만한 부분이 있어서 코멘트 남겼어요.

고민해보시고 재 리뷰요청 주세요! 🙏

> 1. 현재 엔티티 구성에서 다대일 단방향으로 구성된 Section-Station의 관계와 Favorite-Station 관계, Favorite-Member 관계가 있습니다. 이 관계들은 서로 생명주기가 완벽하게 일치하지는 않고 Station, Member에서 Section, Favorite의 정보를 보관하고 있을 필요가 없을 것 같다는 생각에서 이렇게 구성하였습니다! 그런데 Station, Member가 삭제될 때는 삭제되는 Station, Member를 포함하고 있는 Section, Favorite도 삭제되어야 할 것 같은데 여기서 고민이 되는 부분이 있습니다.
> -> StationService, MemberService에서 수행되는 delete 로직에서 삭제되기 전 Section, Favorite을 삭제하는 것
> -> 다대일 단방향 관계를 다대일 양방향으로 수정해 영속성 전이로 관리
> 어떤 방법이 더 낫다고 생각하시는지 궁금합니다!
>
> 좋은 고민 해주셨네요! 👍🏻
> 우선 Section이 Station를 알고있는걸 Section이 Station을 의존한다고 표현해요.
> 하지만 반대로 Station이 Section에 대해 관심을 갖거나 의존해야하는지 고민해볼 필요가 있겠네요.
> StationService에서 Section을 변경하거나 지우는것도 마찬가지입니다.
> 제 생각에는 Station이 삭제될때 Section이 지워져야하는건 Station의 관심사가 아닌 Section의 관심사라고 생각해요!
> Station이 삭제될때 삭제되었다는 이벤트가 생기고 그 이벤트를 Section이 소비하여 삭제처리하면 의존성의 방향을 바꿀수 있어요!
> 레거시코드 리팩터링 미션을 하실때 이 문제에 대해 고민하시게 될거에요! 😀
> 삭제하지 않아도 괜찮습니다. 지금은 참고정도만 해주세요 🙏
> 
> 2. 현재 구현된 즐겨찾기 기능에서 딱히 구간에 대한 검증을 하고 있지는 않은데 source와 target이 서로 연결되었을 경우에만 즐겨찾기를 추가할 수 있도록 하는 validation을 추가하는 것이 좋을까요?
>
> 즐겨찾기에선 source와 target이 연결되었는지에 대한 관심을 가질 필요는 없을것같아요!
> 그렇기 때문에 validation을 추가할 필요는 없다고 생각합니다.
> 저는 애플리케이션의 각 도메인이 서로 관심사를 적절히 분리해야한다고 생각해요.
> 서로가 복잡하게 얽혀있다면 애플리케이션의 복잡도가 매우 올라가고 문제가 생길 가능성이 높아지거나 더이상 유지보수할수 없는 코드가 되어버린다고 생각해요. 😀

jisu1211 committed:
refactor: 3단계 피드백 반영
- 메서드명 수정
- 불필요한 메서드 제거
- @Transactional을 readOnly로 변경

jisu1211 commented:
피드백 주신 내용들 중 궁금한 내용이 추가적으로 생겨서 댓글에 달아두었습니다!
그래서 일단 일부만 피드백 반영하여 리뷰 요청 드렸습니다 ㅎㅎ
나머지는 근환님 의견을 듣고 진행하고자 합니다~ 감사합니다! 😊

geunhwanlee reviewed (changes requested):
리뷰사항 반영 잘 해주셨네요! 👍🏻

질문에 대한 답을 달아두었어요. 😀

한번 고민해보시고 재 리뷰요청 주세요. 🙏

그럼 화이팅하세요! 🔥

jisu1211 committed:
refactor: Favorite 리팩터링
- source와 target이 같을 경우 validation 추가
- 같은 회원인지 비교하는 로직 추가

test: FavoriteTest, MemberTest 수정

jisu1211 committed:
refactor:  FavoriteService 리팩터링
- 내부에서만 사용하는 메서드 @Transactional 제거하고 private 메서드로 변경
- 즐겨찾기 삭제시 로그인한 유저가 즐겨찾기를 생성한 유저인지 확인하는 로직 추가

test: 즐겨찾기 인수 테스트 케이스 추가

jisu1211 committed:
refactor: 내부에서만 사용하는 메서드는 @Transactional을 제거하고 private 메서드로 변경

geunhwanlee commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,75 @@
+    @Transactional
+    public void deleteFavorite(LoginMember loginMember, Long favoriteId) {
+        Favorite favorite = findById(favoriteId);
+        favorite.checkLoginMember(loginMember.getEmail());
```
email보다는 memberId를 비교하는게 어떨까 싶어요 🤔

jisu1211 replied:
넵 다음 단계 진행하면서 수정하겠습니다! 😊

geunhwanlee reviewed (approved):
리뷰사항 반영 잘해주셨네요 👍🏻

충분히 잘해주셔서 머지하도록 하겠습니다.

그럼 다음단계도 화이팅하세요 🔥
